File: Leetcode/16. 3Sum Closest/solution.py
class Solution:
    def threeSumClosest(self, nums: List[int], target: int) -> int:
        nums.sort()
        n = len(nums) - 1
        min_dif = float('inf')
        current_sum = None
        res = -1

        for i in range(len(nums) - 2):
            # if the iteration is the same, continue
            if i != 0 and nums[i] == nums[i-1]:
                continue

            l, r = i + 1, n

            while l < r:
                current_sum = nums[i] + nums[r] + nums[l]
                dif = abs(current_sum - target)

                if(dif < min_dif):
                    min_dif = dif
                    res = current_sum

                if current_sum == target:
                    return current_sum
                if current_sum < target:
                    l += 1
                else:
                    r -= 1

        return res

File: Leetcode/16. 3Sum Closest/solution.rs
impl Solution {
    pub fn three_sum_closest(mut nums: Vec<i32>, target: i32) -> i32 {
        let mut min_dif = i32::MAX;
        let n = nums.len() - 1;
        let mut res = -1;
        
        nums.sort();
        
        for i in 0..nums.len() - 2 {
            
            if i != 0 && nums[i] == nums[i-1] {continue}
            
            let (mut l, mut r) = (i + 1, n);
            while l < r {
                let sum = nums[i] + nums[l] + nums[r];
                let dif = (sum - target).abs();
                
                if dif < min_dif {
                    min_dif = dif;
                    res = sum;
                }
                
                if sum == target {
                    return sum
                }
                
                if sum < target {
                    l += 1;
                } else {
                    r -= 1;
                }
            } 
        }
        res
    }
}